### PR TITLE
[osx/ios] - refactor videosync implementations

### PR DESCRIPTION
**Pull request by Memphiz**
As the title says - this adds proper videosync implementations for ios and osx (and splits up the big mess we had before).

@FernetMenta for review - i think i don't need locking anymore with this - maybe @jingai for review too

### CODE REVIEW

**Comment by jingai** [comment]
Building and testing now.

File: xbmc/video/videosync/VideoSyncOsx.cpp [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+#include <QuartzCore/CVDisplayLink.h>
+#include "osx/CocoaInterface.h"
+
+// Called by the Core Video Display Link whenever it's appropriate to render a frame.
+static CVReturn DisplayLinkCallBack(CVDisplayLinkRef displayLink, const CVTimeStamp* inNow, const CVTimeStamp* inOutputTime, CVOptionFlags flagsIn, CVOptionFlags* flagsOut, void* displayLinkContext)
+{
+  // Create an autorelease pool (necessary to call into non-Obj-C code from Obj-C code)
+  void* pool = Cocoa_Create_AutoReleasePool();
+  
+  CVideoSyncOsx *VideoSyncOsx = reinterpret_cast<CVideoSyncOsx*>(displayLinkContext);4
```
Comment (jingai) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: Typo here -- trailing '4'

**Comment by Memphiz** [comment]
thx typo fixed ...

**Comment by jingai** [comment]
Patches look good to me, and it works as-intended on my machine here, with one caveat:  I applied these against the Helix branch, because master currently isn't upgrading my database properly and I end up with no content (haven't looked into why yet, but this is unrelated anyway).

Just looking at what changed in regards to these areas of the code between Helix and I*******, it should be fine on master too though.

File: xbmc/windowing/osx/WinSystemIOS.mm [hunk 2]
```diff
@@ -346,26 +367,56 @@
...
+//-------------------DisplayLink stuff
+@implementation IOSDisplayLinkCallback
+@synthesize _videoSyncImpl;
+//--------------------------------------------------------------
+- (void) runDisplayLink;
 {
-  if (m_VideoSync)
-    m_VideoSync->VblankHandler(nowtime, fps);
+  NSAutoreleasePool* pool = [[NSAutoreleasePool alloc] init];
+  if (_videoSyncImpl != nil)
```
Comment (FernetMenta) [inline xbmc/windowing/osx/WinSystemIOS.mm]: I am not familiar with IOS specifics. Are you saying that at task switch can't happen here? Is pool a kind of lock?

Comment (Memphiz) [inline xbmc/windowing/osx/WinSystemIOS.mm]: no - my thinking was that invalidate waits for the last callback before returning - mhhh

Comment (FernetMenta) [inline xbmc/windowing/osx/WinSystemIOS.mm]: assume videoSyncImpl != nil, task switch, ref clock clears down -> crash

File: xbmc/video/videosync/VideoSyncDarwin.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  
+  return setupOk;
+}
+
+void CVideoSyncDarwin::Run(volatile bool& stop)
+{
+  //because cocoa has a vblank callback, we just keep sleeping until we're asked to stop the thread
+  while(!stop && !m_abort)
+  {
+    Sleep(1000);
```
Comment (FernetMenta) [inline xbmc/video/videosync/VideoSyncDarwin.cpp]: I would reduce to 100ms. 1 sec is kind of long. a user may observe this delay when stopping playback.

File: xbmc/windowing/osx/WinSystemIOS.mm (hunk 2 above)
Comment (Memphiz) [inline xbmc/windowing/osx/WinSystemIOS.mm]: no my assumption was that deinit doesn't return as long as a callback is running - so refclock can't tear down until deinitdisplaylink returns

Comment (FernetMenta) [inline xbmc/windowing/osx/WinSystemIOS.mm]: > no my assumption was that deinit doesn't return as long as a callback is running

that would be similar behavior as a lock. that's why I asked.

**Comment by Memphiz** [comment]
Now that i look at it i wonder if it would be even better to refactor the osx displaylink stuff into windowing aswell and use the VideoSyncDarwin for both. I think thats cleaner (and less source files). Do you agree from an architecture POV @FernetMenta ?

**Comment by FernetMenta** [comment]
I won't create a dependency to windowing for cocoa. If you look only at OSX without the common base class for IOS and OSX, you see a clean implementation for cocoa. Straightforward and easy to understand. IMO the inheritance hierarchy creates an unnecessary complexity. I would implement the IOS vsync separately from OSX.

**Comment by Memphiz** [comment]
But that would dupe a bunch of code - basically all that is in VideoSyncDarwin now...

**Comment by FernetMenta** [comment]
I see no issue in have the just a couple of line duplicated. VideoSyncDRM also registers into windowing but you won't create a common base class to it and Darwin, right?
If code dupe reduction is the only argument for using polymorphism, it is a good indicator that you shouldn't do so.

**Comment by Memphiz** [comment]
changed and got rid of the darwin base class

File: xbmc/video/videosync/VideoSyncOsx.cpp [hunk 4]
```diff
@@ -20,135 +20,112 @@
...
-  return m_fps;
+// Called by the Core Video Display Link whenever it's appropriate to render a frame.
+static CVReturn DisplayLinkCallBack(CVDisplayLinkRef displayLink, const CVTimeStamp* inNow, const CVTimeStamp* inOutputTime, CVOptionFlags flagsIn, CVOptionFlags* flagsOut, void* displayLinkContext)
+{
+  // Create an autorelease pool (necessary to call into non-Obj-C code from Obj-C code)
+  void* pool = Cocoa_Create_AutoReleasePool();
+  
+  CVideoSyncOsx *VideoSyncOsx = reinterpret_cast<CVideoSyncOsx*>(displayLinkContext);
+  // TODO - is it better to use CurrentHostCounter here?!?
+  VideoSyncOsx->VblankHandler(inOutputTime->hostTime);
```
Comment (FernetMenta) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: how accurate is inOutputTime? Probably not mission critical whether this os CurrentHostCounter is used. CurrentHostCounter would be in line with IOS and DRM

Comment (Memphiz) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: interesting - read:
https://developer.apple.com/library/mac/documentation/QuartzCore/Reference/CVDisplayLinkRef/#//apple_ref/c/tdef/CVDisplayLinkOutputCallback

inNow - A pointer to the current time.
inOutputTime - A pointer to the time that the frame will be displayed.

Comment (FernetMenta) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: hmm, are you sure the callback fires on vertical retrace signal? this reads like inOutputTime - inNow is the time to next vblank.

Comment (Memphiz) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: https://github.com/xbmc/xbmc/pull/6455/files#diff-9ee307fa2e087e837cf2d430925d4d4eR87 <- looks like what you say is done here

Comment (FernetMenta) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: not really. this just counts the vblanks. we assume vertical retrace at the time when UpdateClock is called. the clock is fine adjusted in a way that the discrete time is most accurate in the middle of two vblanks. to that time we sync audio.
can you log both times (and the diff between them) when playing some 24Hz video?

Comment (Memphiz) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: @FernetMenta here you go:

http://pastebin.com/e0AnipHP

Refreshrate adaption turned on and sync to display - playing a 23.976 fps movie

Comment (FernetMenta) [inline xbmc/video/videosync/VideoSyncOsx.cpp]: that looks good. it is approximately frametime. that's the time it takes right after a vblank to display the next frame. perfect. maybe inOutputTime does also observe a skipped vblank -> I would keep this as time.

**Comment by Memphiz** [comment]
removed the todo comment ... should be ready for jenkins now (will test something more and then kick a build)

**Comment by Memphiz** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
looks good

**Pull request merged**